### ISSUE TITLE
Bug: no-unnecessary-condition complains about returning a constant true value from a filter predicate function

### ISSUE THREAD

**Issue by TornOne** [typescript-eslint/typescript-eslint#7143]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=MYewdgzgLgBAhgJwTAvDA2gIjpgNDTAIzwOEwF0BuAWACgATAUwBtGpH4l0AGKu0SLABmAS2bsEjegEEkcAJ6pOCAHSjxjBAAotASlQA+GFAQBXRrspA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
const arr = ["a", "b", "c"];
delete arr[0];
const filteredArray = arr.filter(() => true);
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  },
};
```

### Expected Result

I expected ESLint to not complain about returning an always true value from a filter predicate function.

### Actual Result

ESLint complains about returning an always true value from a filter predicate function.

### Additional Info

`filter(() => true)` doesn't always return the same array it was called on. In case elements from the array have been removed, they are not even passed to the predicate function. Note that this is not the same as the array element being `undefined`, as those values are passed to the predicate function.

**Comment by TornOne** [comment]
I do not have a use case for distinguishing between empty slots and undefined, but strictly speaking, they are not equivalent. It did not cross my mind to type my array as `(string | undefined)[]`, as I was not planning to store any undefined values there.

**Comment by Josh-Cena** [comment]
I would strongly encourage you to write:

```ts
const arr: (string | undefined)[] = ["a", "b", "c"];
arr[0] = undefined;
const filteredArray = arr.filter((x) => x);
```

It's more performant and readable.

**Comment by bradzacher** [comment]
I'm okay with this erroring because almost every time you have a constant condition in your filter - it's an accident.

Sparse arrays are strongly advised against in modern JS because their behaviour (or even their existence) is generally very poorly understood.

If this is something you want in your code - use a disable comment and note why you're using a constant condition to help explain your code to future devs.
